lib/oai.py

```diff
@@ -116,40 +116,71 @@ def __init__(self, root, logger=logging):
         return
     
     def list_sets(self):
-        #e.g. http://dspace.mit.edu/oai/request?verb=ListSets
-        qstr = urllib.urlencode({'verb' : 'ListSets'})
-        url = self.root + '?' + qstr
-        self.logger.debug('OAI request URL: {0}'.format(url))
-        start_t = time.time()
-        try:
-            content = urllib2.urlopen(url).read()
-        except urllib2.URLError as e:
-            raise OAIHTTPError("list_sets could not make request: %s" % \
-                               e.reason)
-        except urllib2.HTTPError as e:
-            raise OAIHTTPError("list_sets got status %d: %s" % \
-                               (e.code, e.reason))
-        retrieved_t = time.time()
-        self.logger.debug('Retrieved in {0}s'.format(retrieved_t - start_t))
+
         sets = []
+        resumptionToken = ''
+        #e.g. http://dspace.mit.edu/oai/request?verb=ListSets
+        params = {'verb' : 'ListSets'}
+
+        while True:
+
+            if resumptionToken:
+                params['resumptionToken'] = resumptionToken
+
+            qstr = urllib.urlencode(params)
+
+            url = self.root + '?' + qstr
+            self.logger.debug('OAI request URL: {0}'.format(url))
+            start_t = time.time()
+            try:
+                content = urllib2.urlopen(url).read()
+            except urllib2.URLError as e:
+                raise OAIHTTPError("list_sets could not make request: %s" % \
+                                   e.reason)
+            except urllib2.HTTPError as e:
+                raise OAIHTTPError("list_sets got status %d: %s" % \
+                                   (e.code, e.reason))
+            retrieved_t = time.time()
+            self.logger.debug('Retrieved in {0}s'.format(retrieved_t - start_t))
+
+            # sets are now passed in
+            # sets = []
```

Comment:
The function would still be clear enough without this comment because `sets` is clearly assigned up above.